waiter/src/waiter/scheduler/cook.clj

```diff
@@ -0,0 +1,523 @@
+            [metrics.timers :as timers]
+            [schema.core :as s]
+            [slingshot.slingshot :as ss]
+            [waiter.mesos.mesos :as mesos]
+            [waiter.metrics :as metrics]
+            [waiter.scheduler :as scheduler]
+            [waiter.util.date-utils :as du]
+            [waiter.util.http-utils :as http-utils]
+            [waiter.util.utils :as utils])
+  (:import (clojure.lang ExceptionInfo)
+           (java.util UUID)))
+
+;; Interactions with Cook API
+
+(defn post-jobs
+  "Post a job description to Cook."
+  [{:keys [http-client impersonate spnego-auth url]} job-description]
+  (let [run-as-user (-> job-description :jobs first :labels :user)]
+    (http-utils/http-request
+      http-client
+      (str url "/jobs")
+      :accept "application/json"
+      :body (json/write-str job-description)
+      :content-type "application/json"
+      :headers (cond-> {} impersonate (assoc "x-cook-impersonate" run-as-user))
+      :request-method :post
+      :spnego-auth spnego-auth)))
+
+(defn delete-jobs
+  "Delete the specified jobs on Cook."
+  [{:keys [http-client impersonate spnego-auth url]} run-as-user job-uuids]
+  (http-utils/http-request
+    http-client
+    (str url "/rawscheduler")
+    :accept "application/json"
+    :content-type "application/json"
+    :headers (cond-> {} impersonate (assoc "x-cook-impersonate" run-as-user))
+    :query-string {:job job-uuids}
+    :request-method :delete
+    :spnego-auth spnego-auth))
+
+(defn get-jobs
+  "Retrieves running or waiting jobs for the specified user on Cook."
+  [{:keys [http-client spnego-auth url]} user states &
+   {:keys [end-time search-interval service-id start-time]
+    :or {search-interval (t/days 7)}}]
+  (let [end-time (or end-time (t/now))
+        end-time-str (du/date-to-str end-time)
+        start-time (or start-time (t/minus end-time search-interval))
+        start-time-str (du/date-to-str start-time)]
+    (http-utils/http-request
+      http-client
+      (str url "/jobs")
+      :accept "application/json"
+      :content-type "application/json"
+      :query-string (cond-> {:end end-time-str
+                             :start start-time-str
+                             :state states
+                             :user user}
+                      service-id (assoc :name (str service-id "*")))
+      :request-method :get
+      :spnego-auth spnego-auth)))
+
+;; Instance health checks
+
+(let [http-client (http-utils/http-client-factory {:conn-timeout 10000
+                                                   :socket-timeout 10000
+                                                   :spnego-auth false})
+      ;; TODO make this cache configurable
+      healthy-instance-cache (-> {}
+                                 (cache/fifo-cache-factory :threshold 5000)
+                                 (cache/ttl-cache-factory :ttl (-> 10 t/seconds t/in-millis))
+                                 atom)]
+  (defn- instance-healthy?
```

Comment:
TODO Move health check out of critical path

Reply:
Added TODO in code.